### PR TITLE
[TR] PIM-3997: rework pim security

### PR DESCRIPTION
**Pull request by sumbobyboys**

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php [hunk 1]
```diff
@@ -18,41 +18,72 @@
...
+        foreach ($filteredAttrGroups as $attributeGroup) {
+            $normalizedAttributes[$attributeGroup->getCode()] = $this->normalizer->normalize($attributeGroup, 'json');
         }
 
         return new JsonResponse($normalizedAttributes);
     }
 
+    /**
+     * get a single attribute group
+     * @param integer $id
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php]: int

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php [hunk 2]
```diff
@@ -18,41 +18,72 @@
...
+     * @return JsonResponse
+     */
     public function getAction($id)
     {
         $attributeGroup = $this->attributeGroupGroupRepo->findOneById($id);
 
-        // if (!$this->securityContext->isGranted(Attributes::VIEW_ATTRIBUTES, $attributeGroup)) {
-        //     throw new AccessDeniedHttpException('You are not authorized to see this attribute group');
-        // }
+        if (null === $attributeGroup ||
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php]: seems like you meant !==

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php [hunk 3]
```diff
@@ -18,41 +18,72 @@
...
+            $this->objectFilter->filterObject($attributeGroup, 'pim:internal_api:attribute_group:view')
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php]: $attributeGroup = $this->objectFilter ....

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php [hunk 4]
```diff
@@ -18,41 +18,72 @@
...
+        ) {
+            throw new NotFoundHttpException(sprintf('No attribute group found for id "%s"', $id));
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php]: "%d"

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeRestController.php [hunk 5]
```diff
@@ -17,33 +19,66 @@
...
-        $normalizedAttributes = $this->normalizer->normalize($attributes, 'json');
+        $filteredAttributes = $this->collectionFilter->filterCollection($attributes, 'pim:internal_api:attribute:view');
+        $normalizedAttributes = $this->normalizer->normalize($filteredAttributes, 'internal_api');
 
         return new JsonResponse($normalizedAttributes);
     }
 
+    /**
+     * Get a single attribute
+     * @param integer $id
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeRestController.php]: int

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js [hunk 6]
```diff
@@ -6,174 +6,175 @@ define([
...
-                    $container.empty();
-                    _.each(elements, _.bind(function (element) {
-                        if (typeof element.render === 'function') {
-                            $container.append(element.render().$el);
-                        } else {
-                            $container.append(element);
-                        }
+                    mediator.trigger('field:extension:add', {'field': this});
+                    _.each(this.elements, _.bind(function (elements, position) {
+                        var $container = this.$('.' + position + '-elements-container');
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js]: we should use a document fragment here at some point

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes/variant-group.js [hunk 7]
```diff
@@ -7,30 +7,35 @@ define(
...
-    function (_, Backbone, BaseForm, FieldManager, VariantGroupManager, variantGroupTemplate) {
+    function (_, Backbone, BaseForm, FieldManager, VariantGroupManager, mediator, variantGroupTemplate) {
         return BaseForm.extend({
             template: _.template(variantGroupTemplate),
-            render: function () {
+            configure: function() {
+                mediator.on('field:extension:add', _.bind(this.addExtension, this));
+
+                return $.when(
+                    BaseForm.prototype.configure.apply(this, arguments)
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes/variant-group.js]: just return the result from the parent method call

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/save.js [hunk 8]
```diff
@@ -34,7 +34,7 @@ define(
                 return this;
             },
             save: function () {
-                var product   = this.getData();
+                var product = $.extend(true, {}, this.getData());
```
Comment (filipsalpe) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/save.js]: we should swap underscore for lodash so we can do _.cloneDeep :smiling_imp:

File: src/Pim/Bundle/TransformBundle/Normalizer/Structured/FamilyNormalizer.php [hunk 9]
```diff
@@ -26,28 +26,22 @@ class FamilyNormalizer implements NormalizerInterface
...
      */
-    protected $attributeFilters = [];
+    protected $collectionFilter;
 
     /**
      * Constructor
      *
-     * @param TranslationNormalizer $transNormalizer
+     * @param TranslationNormalizer     $transNormalizer
+     * @param CollectionFilterInterface $collectionFilter
```
Comment (filipsalpe) [inline src/Pim/Bundle/TransformBundle/Normalizer/Structured/FamilyNormalizer.php]: CollectionFilterInterface|null

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php (hunk 3 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php]: Nope, we only test if the attribute group should be filtered

File: src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php (hunk 2 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Controller/AttributeGroupRestController.php]: fixed

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js (hunk 6 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js]: yeah, definitly

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/save.js (hunk 8 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/save.js]: ^^yeah :)

**Comment by filipsalpe** [comment]
GTM once travis is green

**Pull request merged**